### ISSUE TITLE
“ERROR:  variable not found in subplan target list” with PostgreSQL 11.15, 12.10, 13.6

### ISSUE THREAD

**Issue by andersk** [pgroonga/pgroonga#203]
With today’s release of PostgreSQL 11.15, 12.10, and 13.6, [Zulip](https://github.com/zulip/zulip/)’s automated test suite started nondeterministically failing. I reduced the problem to the following, which can be reproduced from a freshly created database (here using PostgreSQL 11.15 with PGroonga 2.3.4 on Debian 10):

```
psql (11.15 (Debian 11.15-1.pgdg100+1))
Type "help" for help.

test=# CREATE EXTENSION pgroonga;
CREATE EXTENSION
test=# CREATE TABLE t AS SELECT CAST(c AS text) FROM generate_series(1, 10000) AS c;
SELECT 10000
test=# CREATE INDEX t_c ON t USING pgroonga (c);
CREATE INDEX
test=# VACUUM t;
VACUUM
test=# SELECT COUNT(*) FROM t;
ERROR:  variable not found in subplan target list
```

The problem disappears if I downgrade PostgreSQL to 11.14, and reappears if I upgrade back to 11.15. I confirmed by bisecting PostgreSQL that the problem first appears with postgres/postgres@ec367452179c9b62e8d578c9abfab953838fb01d “Fix index-only scan plans, take 2.”

Possibly related:

- #101
- [PostgreSQL #16303: A condtion whether an index-only scan is possible includes a wrong](https://www.postgresql.org/message-id/flat/16303-889581eb40fc0691%40postgresql.org)

[Edited to clean up the confusion from before I figured out how to reliably reproduce.]

**Comment by andersk** [comment]
Confirmed that the symptom appears with postgres/postgres@ec367452179c9b62e8d578c9abfab953838fb01d and does not appear with its parent postgres/postgres@79a4bf45e83ec92ea24421c0ce0790ecb0aa1542.

Query plan with PostgreSQL 11.14 (working):

```psql
zulip_test_template_9623815_10=> EXPLAIN SELECT COUNT(*) FROM zerver_message;
                                                   QUERY PLAN                                                    
-----------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=0.38..0.39 rows=1 width=8)
   ->  Index Only Scan using zerver_message_search_pgroonga on zerver_message  (cost=0.00..0.30 rows=30 width=0)
(2 rows)
```

After `VACUUM FULL`, the plan stops using the index, which probably explains why that makes the symptom disappear in 11.15.

```psql
zulip_test_template_9623815_10=> VACUUM FULL;
VACUUM
zulip_test_template_9623815_10=> EXPLAIN SELECT COUNT(*) FROM zerver_message;
                             QUERY PLAN                              
---------------------------------------------------------------------
 Aggregate  (cost=1.38..1.39 rows=1 width=8)
   ->  Seq Scan on zerver_message  (cost=0.00..1.30 rows=30 width=0)
(2 rows)
```

Possibly related:

* #101

**Comment by andersk** [comment]
This seems to reliably reproduce the problem in PostgreSQL 11.15 with PGroonga 2.3.4, starting from a freshly created database:

```sql
CREATE EXTENSION IF NOT EXISTS pgroonga;
DROP TABLE IF EXISTS t;
CREATE TABLE t AS SELECT c FROM generate_series(1, 10000) AS c;
CREATE INDEX t_c ON t USING pgroonga (c);
VACUUM t;
SELECT COUNT(*) FROM t;
```

**Comment by kou** [comment]
Thanks for investigating this!

Our CI is also broken with new PostgreSQL: https://github.com/pgroonga/pgroonga/runs/5149762901?check_suite_focus=true

https://www.postgresql.org/message-id/602391641208390%40iva4-92c901fae84c.qloud-c.yandex.net says partial-retrieval index-only scan but our case is non-retrievable index-only scan. In non-retrievable index-only scan, the error is occurred.

We asked about non-retrievable index-only scan on the PostgreSQL mailing list in the past: https://www.postgresql.org/message-id/5148.1584372043%40sss.pgh.pa.us
We thought non-retrievable index-only scan should not be used but PostgreSQL may use it as a valid plan. So I think that our case should be supported with postgres/postgres@ec367452179c9b62e8d578c9abfab953838fb01d “Fix index-only scan plans, take 2.”

Could you ask about this case on the PostgreSQL mailing list https://www.postgresql.org/list/pgsql-hackers/ ?

The following patch fixes our case and PostgreSQL's test cases are still passed but a review by the original author is needed:

```diff
--- postgresql-11.15.orig/src/backend/optimizer/plan/setrefs.c	2022-02-08 06:20:23.000000000 +0900
+++ postgresql-11.15/src/backend/optimizer/plan/setrefs.c	2022-02-12 07:32:20.355567317 +0900
@@ -1034,7 +1034,7 @@
 	{
 		TargetEntry *indextle = (TargetEntry *) lfirst(lc);
 
-		if (!indextle->resjunk)
+		if (!indextle->resjunk || indextle->expr->type == T_Var)
 			stripped_indextlist = lappend(stripped_indextlist, indextle);
 	}
 
```

**Comment by andersk** [comment]
> Could you ask about this case on the PostgreSQL mailing list?

https://www.postgresql.org/message-id/flat/4c347490-d734-5fdd-d613-1327601b4e7e%40mit.edu

**Comment by kou** [comment]
Thanks!

**Comment by kou** [comment]
This is a PostgreSQL's problem.
The next PostgreSQL release will fix this.

**Comment by fungcheung** [comment]
Sorry to interrupt you, but I need help urgently.

I am using postgis 3.2 extension on postgresql 13.6 and hit the error below
sql> select count(*) from had_burial_poly;
ERROR:  variable not found in subplan target list

Do you know whether it is a bug and will be fixed in next minor version? Any fix in the meantime? Thanks.

**Comment by kou** [comment]
You can use PostgreSQL 13.5 as a workaround.

**Comment by fungcheung** [comment]
I am using below command to get the latest postgresql 13.6 version. How can I get 13.5 and then perform the downgrade? Thank you very much.

sudo yum -y install https://download.postgresql.org/pub/repos/yum/reporpms/EL-7-x86_64/pgdg-redhat-repo-latest.noarch.rpm

**Comment by kou** [comment]
`yum install -y postgresql13-13.5-1PGDG.rhel7.x86_64`

**Comment by fungcheung** [comment]
do I need to uninstall the 13.6 version first?

**Comment by fungcheung** [comment]
is there a fix or db setting we can apply to 13.6 to fix this problem? sorry for too many questions. the app needs to go production very soon, but just find out this problem!

**Comment by kou** [comment]
> do I need to uninstall the 13.6 version first?

I think that you don't need but you can try it.

> is there a fix or db setting we can apply to 13.6 to fix this problem?

I think that there is no workaround with 13.6.

You can ask about this on https://www.postgresql.org/list/pgsql-general/ . Here is not the right place to ask about this.

**Comment by icyerasor** [comment]
> is there a fix or db setting we can apply to 13.6 to fix this problem? sorry for too many questions. the app needs to go production very soon, but just find out this problem!

Doing a `VACUUM FULL;` seems to have fixed the issue for me (temporarily). You need to upgrade to a newer version of Postgresql though to resolve it completely I guess.
